pkg/logs/sds/scanner.go

```diff
@@ -0,0 +1,295 @@
+)
+
+const ScannedTag = "sds_agent:true"
+
+const (
+	tlmSDSConfigurationError   = "datadog.agent.sds.configuration.error"
+	tlmSDSConfigurationSuccess = "datadog.agent.sds.configuration.success"
+	tlmSDSConfiguredRules      = "datadog.agent.sds.rules"
+)
+
+// Scanner wraps an SDS Scanner implementation, adds reconfiguration
+// capabilities and telemetry on top of it.
+// Most of Scanner methods are not thread safe for performance reasons, the caller
+// has to ensure of the thread safety.
+type Scanner struct {
+	*sds.Scanner
+	sync.Mutex
+
+	standardRules StandardRulesConfig
+	// rawConfig is the raw config previously received through RC.
+	rawConfig []byte
+	// configuredRules are stored on configuration to retrieve rules
+	// information on match. Use this read-only.
+	configuredRules []RuleConfig
+}
+
+// CreateScanner creates an SDS scanner.
+// Use `Reconfigure` to configure it manually.
+func CreateScanner() *Scanner {
+	scanner := &Scanner{}
+	log.Debugf("creating a new SDS scanner (internal id: %p)", scanner)
+	return scanner
+}
+
+// Reconfigure uses the given `ReconfigureOrder` to reconfigure in-memory
+// standard rules or user configuration.
+// The order contains both the kind of reconfiguration to do and the raw bytes
+// to apply the reconfiguration.
+// When receiving standard rules, user configuration are reloaded and scanners are
+// recreated to use the newly received standard rules.
+// This method is thread safe, a scan can't happen at the same time.
+func (s *Scanner) Reconfigure(order ReconfigureOrder) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s == nil {
+		log.Warn("Trying to reconfigure a nil Scanner")
+		return nil
+	}
+
+	log.Debugf("Reconfiguring SDS scanner (internal id: %p)", s)
+
+	switch order.Type {
+	case StandardRules:
+		err := s.reconfigureStandardRules(order.Config)
+		// if we already received a configuration,
+		// reapply it now that the standard rules have changed.
+		if s.rawConfig != nil {
+			if rerr := s.reconfigureRules(s.rawConfig); rerr != nil {
+				log.Error("Can't reconfigure SDS after having received standard rules:", rerr)
+				s.rawConfig = nil // we drop this configuration because it is unusable
+				if err == nil {
+					err = rerr
+				}
+			}
+		}
+		return err
+	case AgentConfig:
+		return s.reconfigureRules(order.Config)
+	}
+
+	return fmt.Errorf("Scanner.Reconfigure: Unknown order type: %v", order.Type)
+}
+
+// reconfigureStandardRules stores in-memory standard rules received through RC.
+// This is NOT reconfiguring the internal SDS scanner, call `reconfigureRules`
+// if you have to.
+// This method is NOT thread safe, the caller has to ensure the thread safety.
+func (s *Scanner) reconfigureStandardRules(rawConfig []byte) error {
+	if rawConfig == nil {
+		return fmt.Errorf("Invalid nil raw configuration for standard rules")
+	}
+
+	var standardRules StandardRulesConfig
+	if err := json.Unmarshal(rawConfig, &standardRules); err != nil {
```

Comment:
It's a good idea, l'll look into this. (the number of pipelines is the number of logs agent pipelines -> 4 by default)

Reply:
Unfortunately I can't with current design: the types are not accessible before propagating to pipelines (it would create an import cycle).
If you don't mind, I'll keep the unmarshal per pipeline for now, as it is only on reconfiguration, it should be OK.